pkg/server/plugin/noderesolver/azure/msi.go

```diff
@@ -0,0 +1,344 @@
+	reVirtualNetworkSubnetID = regexp.MustCompile(`^/subscriptions/[^/]+/resourceGroups/([^/]+)/providers/Microsoft.Network/virtualNetworks/([^/]+)/subnets/([^/]+)$`)
+)
+
+type TenantConfig struct {
+	SubscriptionId string `hcl:"subscription_id" json:"subscription_id"`
+	AppId          string `hcl:"app_id" json:"app_id"`
+	AppSecret      string `hcl:"app_secret" json:"app_secret"`
+}
+
+type MSIResolverConfig struct {
+	UseMSI  bool                    `hcl:"use_msi" json:"use_msi"`
+	Tenants map[string]TenantConfig `hcl:"tenants" json:"tenants"`
+}
+
+type MSIResolverPlugin struct {
+	mu            sync.RWMutex
+	msiClient     apiClient
+	tenantClients map[string]apiClient
+
+	hooks struct {
+		newClient             func(string, autorest.Authorizer) apiClient
+		fetchInstanceMetadata func(context.Context, azure.HTTPClient) (*azure.InstanceMetadata, error)
+	}
+}
+
+var _ noderesolver.Plugin = (*MSIResolverPlugin)(nil)
+
+func NewMSIResolverPlugin() *MSIResolverPlugin {
+	p := &MSIResolverPlugin{}
+	p.hooks.newClient = newAzureClient
+	p.hooks.fetchInstanceMetadata = azure.FetchInstanceMetadata
+	return p
+}
+
+func (p *MSIResolverPlugin) Resolve(ctx context.Context, req *noderesolver.ResolveRequest) (*noderesolver.ResolveResponse, error) {
+	resp := &noderesolver.ResolveResponse{
+		Map: make(map[string]*common.Selectors),
+	}
+	for _, spiffeID := range req.BaseSpiffeIdList {
+		selectors, err := p.resolveSpiffeID(ctx, spiffeID)
+		if err != nil {
+			return nil, err
+		}
+		if selectors != nil {
+			resp.Map[spiffeID] = selectors
+		}
+	}
+
+	return resp, nil
+}
+
+func (p *MSIResolverPlugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
+	config := new(MSIResolverConfig)
+	if err := hcl.Decode(config, req.Configuration); err != nil {
+		return nil, msiError.New("unable to decode configuration: %v", err)
+	}
+
+	var msiClient apiClient
+	var tenantClients map[string]apiClient
+
+	if config.UseMSI {
+		if len(config.Tenants) > 0 {
+			return nil, msiError.New("configuration cannot have tenants when using MSI")
+		}
+		instanceMetadata, err := p.hooks.fetchInstanceMetadata(ctx, http.DefaultClient)
+		if err != nil {
+			return nil, msiError.Wrap(err)
+		}
+		authorizer, err := auth.NewMSIConfig().Authorizer()
+		if err != nil {
+			return nil, msiError.Wrap(err)
+		}
+		msiClient = p.hooks.newClient(instanceMetadata.Compute.SubscriptionID, authorizer)
+	} else {
+		if len(config.Tenants) == 0 {
+			return nil, msiError.New("configuration must have at least one tenant when not using MSI")
+		}
+		tenantClients = make(map[string]apiClient)
+		for tenantID, tenant := range config.Tenants {
+			if tenant.SubscriptionId == "" {
+				return nil, msiError.New("misconfigured tenant %q: missing subscription id", tenantID)
+			}
+			if tenant.AppId == "" {
+				return nil, msiError.New("misconfigured tenant %q: missing app id", tenantID)
+			}
+			if tenant.AppSecret == "" {
```

Comment:
If there is an empty value for the app secret, I suggest reading it from an environment variable.

Reply:
:+1:

Reply:
nevermind.... see my other comment.